blog/2022-11-21-golang-memory-dump.md

```diff
@@ -0,0 +1,179 @@
+   25.59GB 92.43% 92.43%    25.59GB 92.43%  math/big.nat.make (inline)
+         0     0% 92.43%    25.41GB 91.78%  github.com/polarsignals/pprof-example-app-go/fib.Fibonacci
+         0     0% 92.43%    25.41GB 91.78%  math/big.(*Int).Add
+         0     0% 92.43%    25.41GB 91.78%  math/big.nat.add
+    2.02GB  7.30% 99.73%     2.02GB  7.30%  main.allocMem
+    0.02GB 0.073% 99.80%     0.25GB  0.91%  fmt.Sprintln
+         0     0% 99.80%     0.25GB  0.91%  log.Println
+         0     0% 99.80%     0.23GB  0.84%  fmt.(*pp).doPrintln
+         0     0% 99.80%     0.23GB  0.84%  fmt.(*pp).handleMethods
+```
+
+`top 10 -flat` will show you the resource consumption it takes to call a function excluding all function it calls
+
+```pprof
+Active filters:
+   ignore=flat
+Showing nodes accounting for 27.66GB, 99.90% of 27.69GB total
+Dropped 10 nodes (cum <= 0.14GB)
+Showing top 10 nodes out of 17
+      flat  flat%   sum%        cum   cum%
+   25.59GB 92.43% 92.43%    25.59GB 92.43%  math/big.nat.make (inline)
+    2.02GB  7.30% 99.73%     2.02GB  7.30%  main.allocMem
+    0.03GB   0.1% 99.83%     0.21GB  0.76%  math/big.nat.itoa
+    0.02GB 0.073% 99.90%     0.25GB  0.91%  fmt.Sprintln
+         0     0% 99.90%     0.23GB  0.84%  fmt.(*pp).doPrintln
+         0     0% 99.90%     0.23GB  0.84%  fmt.(*pp).handleMethods
+         0     0% 99.90%     0.23GB  0.84%  fmt.(*pp).printArg
+         0     0% 99.90%    25.41GB 91.78%  github.com/polarsignals/pprof-example-app-go/fib.Fibonacci
+         0     0% 99.90%     0.25GB  0.91%  log.Println
+         0     0% 99.90%    25.66GB 92.69%  main.calculateFib
+```
+
+By looking at the output we can see that it's `math/big.nat.make` that takes the most resources.
+
+## Cleanup
+
+To cleanup the resources we created run:
+
+```shell
+kubectl delete -f https://raw.githubusercontent.com/polarsignals/pprof-example-app-go/main/manifests/deployment.yaml
+kubectl delete svc pprof-example-app-go
+kubectl delete pod curl
+```
+
+## Conclusion
+
+The `kubectl debug` command is extremely useful when you want to debug your application and you
+don't have access to a shell or the tools that you need in your normal container.
+
+It saves us from having to install unneeded applications in our container which lowers the amount of potential CVE:s and the time it takes to start your container by lower container size.
+
+Kubectl debug isn't perfect and it won't work for all your uses cases especially since you can't
+attach volumes but it's a great start.
+
+If you really don't want to use a tool like pprof inside a container you could of course write some small application that performs a request to an endpoint and streams the output to an object store or similar without ever storing the data on disk.
+
+When it comes to continues profiling it's probably better to look at tool specifically written for it like [Grafana Phlare](https://grafana.com/oss/phlare/) or [Parca](https://www.parca.dev/docs/overview).
```

Comment:
I haven't spoken about continues profiling at all in the text. I should probably remove it or I should talk a bit about it in the text.

Reply:
I think you should keep it, when reading through it I thought about how to let Datadog monitor this.

Reply:
I think it is good to mention this. It is always good to establish what best practice is.

Reply:
Sadly I don't think datadog got support for it so far, but It will probably come in the future :)
Thanks then I will keep it.